### PR TITLE
Add Atlassian provider

### PR DESCRIPTION
**Pull request by andor0**
Add Atlassian provider

## Description

I’m working on Atlassian oauth. It can be currrently considered WIP. It’s not currently tested as we’re still waiting for app approval from the Atlassian’s side and test suite is not impemented in this branch yet.

## How Has This Been Tested?

Manually. Test suite is not impemented in this branch yet.

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by NickMeves** [comment]
Hey there,

At the moment, the Provider interface and existing providers are undergoing a bit of an overhaul. I don't want to merge any new ones until that wraps up to avoid rework & refactoring on them.

Here's the existing pertinent discussions going on around it:
https://github.com/oauth2-proxy/oauth2-proxy/issues/956#issuecomment-746749329
https://github.com/oauth2-proxy/oauth2-proxy/pull/899

**Comment by andor0** [comment]
> This pull request has been inactive for 60 days. If the pull request is still relevant please comment to re-activate the pull request. If no action is taken within 7 days, the pull request will be marked closed.

We will plan to continue work on this PR.

**Comment by NickMeves** [comment]
This is merging soon which should unblock this: https://github.com/oauth2-proxy/oauth2-proxy/pull/947

File: providers/atlassian.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+func (p *AtlassianProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	err := p.validateUserGroups(ctx, s)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// ValidateSessionState validates the AccessToken
+func (p *AtlassianProvider) ValidateSession(ctx context.Context, s *sessions.SessionState) bool {
```
Comment (NickMeves) [inline providers/atlassian.go:124]: Are you able to implement the `RefreshSessionIfNeeded` (refresh token refresh call to redeem endpoint)?

I'm working on refactoring `RefreshSession` here: https://github.com/oauth2-proxy/oauth2-proxy/pull/1086

I want to make sure our provider implementations stop skipping implementing that if possible. Users who use the `--cookie-refresh` option would be surprised when access tokens aren't refreshed like expected.

There's also a nasty performance bug at the moment when `ValidateSession` is implemented and `RefreshSessionIfNeeded` is not implemented and a `--cookie-refresh` is set that the validate endpoint will be hit on every request after the refresh timer is up.

File: providers/atlassian.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+func makeAtlassianHeader(accessToken string) http.Header {
+	// extra headers required by the Atlassian API when making authenticated requests
+	extraHeaders := map[string]string{
+		acceptHeader: acceptApplicationJSON,
+	}
+	return makeAuthorizationHeader(tokenTypeBearer, accessToken, extraHeaders)
+}
+
+// SetGroups sets Atlassian Groups
+func (p *AtlassianProvider) SetGroups(groups []string) {
```
Comment (NickMeves) [inline providers/atlassian.go:110]: This is all standardizing to using the `AllowedGroups` set on the top-level `ProviderData` struct and using `allowed-groups` in configurations across providers to set the authZ groups.

You can then either use the default `Authorize` interface method from `ProviderData` (which does generic group set membership checks) or override it if you need custom logic (look to the Gitlab Provider to see a sample of this and how it manages multiple group types in the groups set together with prefixes).

File: providers/atlassian.go [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+	p.Groups = groups
+}
+
+// EnrichSessionState updates the Email and User after the initial Redeem
+func (p *AtlassianProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	err := p.validateUserGroups(ctx, s)
```
Comment (NickMeves) [inline providers/atlassian.go:116]: This should only set groups to the `session.Groups` field (so they get passed to upstreams in `X-Forwarded-Groups` header).

`Authorize` is where group authorization happens (It does a set membership check from the user session against the allowed groups on every request).

File: providers/atlassian.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+		redeemURL:   atlassianDefaultRedeemURL,
+		profileURL:  nil,
+		validateURL: atlassianDefaultValidateURL,
+		scope:       atlassianDefaultScope,
+	})
+	p.Prompt = atlassianPrompt
+	return &AtlassianProvider{ProviderData: p}
+}
+
+func makeAtlassianHeader(accessToken string) http.Header {
```
Comment (NickMeves) [inline providers/atlassian.go:101]: This completely duplicates `makeOIDCHeader` in util.go -- lets use that.

If the name is misleading, we can rename to something like `makeAcceptJSONHeader` if you prefer.

**Review by NickMeves — CHANGES_REQUESTED** [review]
Overall this looks very clean! 👍 

Let me know if you have questions about my notes around authorization design or refreshing.

Otherwise, ping me once you get the test suite added and I'll take another look. If you are comfortable using it, we are moving more things to Ginkgo - you can look at `gitlab_test.go` or `keycloak_test.go` for samples in the provider package.

**Comment by ilyakooo0** [comment]
This PR is still relevant.

**Comment by ilyakooo0** [comment]
Still relevant (:

**Pull request closed without merging**